engine/src/resource/product.h

```diff
@@ -27,24 +27,15 @@
 
 #include "resource.h"
 
-class Product
+struct Product
```

Comment:
I think this is better as a class, with its fields `protected` or even `private`. Prefer not to expose internals to other parts of the code except when necessary.

Reply:
I don't actually remember where I found the recommendation to use struct. I used struct as the fields are public and there is no reason right now to hide them. If it becomes necessary, we can switch.
We're paying a cost in readability here. A lot of the old code is harder to read and for no good reason. It isn't doing something fancy or complicated. It's just written that way.

Reply:
I'm mixed on this one. If a structure only has data, then the `struct` is 100% the right call. If there's associated methods, then it should be a class and the data should be more hidden by default with proper getters/setters.

I do generally agree with @stephengtuggy that the data should be protected.

That said, the old code has some issues:
1. we shouldn't be friending if we can help it - just implement the functions directly or wrappers to do so, especially with operators.
2. we shouldn't friend other class objects; those should adhere to the public interface and if necessary improve the public interface.

It may be that we need to open up to general `struct` for a while in order to fix things up before closing them back down after we fix a bunch of relationships, etc - if that's the case, I'm okay with making this kind of change.

Reply:
NOTE: Another option if you want to keep it a `struct` is having non-associated methods - straight functional/free-functions instead of objects.

Reply:
I think it makes sense to use friend functionality where it makes sense. Carrier/Cargo and Store/Product are two examples that come to mind. What do you think?

Reply:
@royfalk Yeah, I think in certain cases using `friend` functionality is OK. That would probably include the cases you mentioned. At least if it turns out to be necessary.